### PR TITLE
Jira/dev 41 sign transaction instruction cont

### PR DESCRIPTION
**Pull request by Kazopl**
## Summary by CodeRabbit

- **New Features**
    - Introduced new parameters across various models to enhance specificity and functionality, including `CertificateType`, `CredentialType`, `DRepType`, `PoolKeyType`, `PoolOwnerType`, `RelayType`, `PoolRewardAccountType`, and `RequiredSignerType`.
    - Added new serialization methods for improved data handling.
    - Enhanced transaction operation capabilities with significant updates.
- **Refactor**
    - Updated and added new constants for better code clarity and management.
    - Refined model properties and factory constructors for increased precision in data representation.
- **Style**
    - Renamed `SingleHostHostname` to `SingleHostName` for consistency.
- **Chores**
    - Added necessary import statements to support new features and refactors.

### CODE REVIEW

File: lib/src/models/parsed_certificate.dart [hunk 1]
```diff
@@ -11,66 +12,78 @@ sealed class ParsedCertificate with _$ParsedCertificate {
   const ParsedCertificate._();
 
   const factory ParsedCertificate.stakeRegistration({
+    required CertificateType type,
```
Comment (AlexDochioiu) [inline lib/src/models/parsed_certificate.dart:15]: Aici nu e bine. Pentru ca tipul certificatului este de fapt constant pentru fiecare factory type.

Cu implementarea sugerata de voi pot crea:
```dart
final stakeCert = ParsedCertificate.stakeRegistration(
  type: CertificateType.stakeDeregistration,
  stakeCredential: ...,
);
```

care ar fi clar incorect (si un caz care nu trebuie sa existe in realitate)

Ce trebuie sa faceti e sa:
1) Stergeti acest `required CertificateType type` pe care l-ati adaugat la toate factory-urile
2) Adaugati la final un field:

```dart
late final type = switch(this) {
  StakeRegistration() => CertificateType.stakeRegistration,
  StakeDeregistration() => CertificateType.stakeDeregistration,
  ...
}
```

Comment (AlexDochioiu) [reply lib/src/models/parsed_certificate.dart:15]: Am observat acum ca aceeasi greseala exista pentru `StakingDataSource`. Si acolo trebuie corectat. Cautati va rog daca mai exista in alta parte aceeasi greseala.

Comment (AlexDochioiu) [reply lib/src/models/parsed_certificate.dart:15]: Cred ca astea sunt toate cazurile in care faceti asta: 
`ParsedCertificate`, `ParsedCVoteDelegation`, `ParsedCredential`, `ParsedDRep`
, `ParsedOutputDestination`, `ParsedPoolKey`, `ParsedPoolOwner`, `ParsedPoolRelay`, `ParsedPoolRewardAccount`, `ParsedRequiredSigner`, `ParsedTxAuxiliaryData`, `SpendingDataSource`, `StakingDataSource`,

File: lib/src/utils/serialization_utils.dart [hunk 2]
```diff
@@ -128,6 +137,20 @@ class SerializationUtils {
     });
   }
 
+  static Uint8List serializeCoin(String coin) {
+    return useBinaryWriter((ByteDataWriter writer) {
+      writeSerializedUint64(writer, BigInt.parse(coin));
+      return writer.toBytes();
+    });
+  }
```
Comment (AlexDochioiu) [inline lib/src/utils/serialization_utils.dart:145]: Cateva chestii trebuie schimbate aici:
a) `coin` ar trebui sa fie `BigInt` nu `String` din start
b) `Uint64` nu e destul. Valoare pentru coin nu are limita. Deci trebuie serializat fara sa limitam valoarea la uint64.

File: lib/src/utils/serialization_utils.dart [hunk 3]
```diff
@@ -128,6 +137,20 @@ class SerializationUtils {
...
+
+  static Uint8List serializeTxTtl(String ttl) {
+    return useBinaryWriter((ByteDataWriter writer) {
+      writeSerializedUint64(writer, BigInt.parse(ttl));
+      return writer.toBytes();
+    });
+  }
```
Comment (AlexDochioiu) [inline lib/src/utils/serialization_utils.dart:152]: ttl ar trebui sa fie de asemenea BigInt. Si la fel ar trebui ceva functie care scrie bigint fara sa faca limita la Uint64

File: lib/src/models/parsed_certificate.dart [hunk 4]
```diff
@@ -11,66 +12,78 @@ sealed class ParsedCertificate with _$ParsedCertificate {
...
 
   const factory ParsedCertificate.stakePoolRegistration({
+    required CertificateType type,
     required ParsedPoolParams pool,
   }) = StakePoolRegistration;
 
   const factory ParsedCertificate.stakePoolRetirement({
+    required CertificateType type,
     required List<int> path,
     required String retirementEpoch,
```
Comment (AlexDochioiu) [inline lib/src/models/parsed_certificate.dart:87]: Please change this to BigInt instead.

File: lib/src/models/parsed_certificate.dart [hunk 5]
```diff
@@ -11,66 +12,78 @@ sealed class ParsedCertificate with _$ParsedCertificate {
...
+    required CertificateType type,
     required ParsedCredential coldCredential,
-     ParsedAnchor? anchor,
+    required ParsedAnchor anchor,
   }) = ResignCommitteeCold;
 
   const factory ParsedCertificate.dRepRegistration({
+    required CertificateType type,
     required ParsedCredential dRepCredential,
     required String deposit,
```
Comment (AlexDochioiu) [inline lib/src/models/parsed_certificate.dart:63]: Please change this to BigInt instead.

File: lib/src/models/parsed_certificate.dart [hunk 6]
```diff
@@ -11,66 +12,78 @@ sealed class ParsedCertificate with _$ParsedCertificate {
...
-     ParsedAnchor? anchor,
+    required ParsedAnchor anchor,
   }) = DRepRegistration;
 
   const factory ParsedCertificate.dRepDeregistration({
+    required CertificateType type,
     required ParsedCredential dRepCredential,
     required String deposit,
```
Comment (AlexDochioiu) [inline lib/src/models/parsed_certificate.dart:70]: Please change this to BigInt instead.

File: lib/src/models/parsed_certificate.dart [hunk 7]
```diff
@@ -11,66 +12,78 @@ sealed class ParsedCertificate with _$ParsedCertificate {
...
 
   const factory ParsedCertificate.stakeDeregistration({
+    required CertificateType type,
     required ParsedCredential stakeCredential,
   }) = StakeDeregistration;
 
   const factory ParsedCertificate.stakeDeregistrationConway({
+    required CertificateType type,
     required ParsedCredential stakeCredential,
     required String deposit,
```
Comment (AlexDochioiu) [inline lib/src/models/parsed_certificate.dart:33]: Please change this to BigInt instead.

File: lib/src/models/parsed_certificate.dart [hunk 8]
```diff
@@ -11,66 +12,78 @@ sealed class ParsedCertificate with _$ParsedCertificate {
...
 
   const factory ParsedCertificate.stakeRegistration({
+    required CertificateType type,
     required ParsedCredential stakeCredential,
   }) = StakeRegistration;
 
   const factory ParsedCertificate.stakeRegistrationConway({
+    required CertificateType type,
     required ParsedCredential stakeCredential,
     required String deposit,
```
Comment (AlexDochioiu) [inline lib/src/models/parsed_certificate.dart:22]: Please change this to BigInt instead.

File: lib/src/utils/serialization_utils.dart [hunk 9]
```diff
@@ -340,4 +363,386 @@ class SerializationUtils {
...
+      }
+      return useBinaryWriter((ByteDataWriter writer) {
+        writer.write(serializeCoin(withdrawal.amount));
+        writerSerializedPath(writer, withdrawalStakeCredential.path);
+        return writer.toBytes();
+      });
+    }
+  }
+
+  static Uint8List serializeTxValidityStart(String validityIntervalStart) {
```
Comment (AlexDochioiu) [inline lib/src/utils/serialization_utils.dart:700]: Please change this to BigInt instead.

File: lib/src/utils/serialization_utils.dart [hunk 10]
```diff
@@ -340,4 +363,386 @@ class SerializationUtils {
...
+
+  static Uint8List serializeAssetGroup(ParsedAssetGroup assetGroup) {
+    return useBinaryWriter((ByteDataWriter writer) {
+      writeSerializedHex(writer, assetGroup.policyIdHex);
+      writer.writeUint32(assetGroup.tokens.length);
+      return writer.toBytes();
+    });
+  }
+
+  static Uint8List serializeToken<T>(ParsedToken token, Uint8List Function(T) serializeTokenAmountFn) {
```
Comment (AlexDochioiu) [inline lib/src/utils/serialization_utils.dart:722]: I am a bit confused why `serializeTokenAmountFn` is passed (and why it is  a function). Maybe I am wrong but I expect this function should always be the same, which means it shouldn't be passed as argument.

File: lib/src/utils/utilities.dart [hunk 11]
```diff
@@ -1,8 +1,19 @@
...
 import 'package:ledger_flutter/ledger_flutter.dart';
 
 Uint8List useBinaryWriter(Uint8List Function(ByteDataWriter writer) invoker) {
   final writer = ByteDataWriter();
   return invoker(writer);
 }
+
+Uint8List ipStringToBytes(String ipString) {
+  try {
+    InternetAddress address = InternetAddress(ipString);
```
Comment (AlexDochioiu) [inline lib/src/utils/utilities.dart:14]: Can you please check if you can find some other way to do this without using something from `dart:io`? Maybe finding some other library and copying the code from there?

P.s. dart:io import doesn't work for flutter web

File: lib/src/utils/serialization_utils.dart [hunk 12]
```diff
@@ -340,4 +363,386 @@ class SerializationUtils {
...
+        writer.write(serializeCoin(withdrawal.amount));
+        writerSerializedPath(writer, withdrawalStakeCredential.path);
+        return writer.toBytes();
+      });
+    }
+  }
+
+  static Uint8List serializeTxValidityStart(String validityIntervalStart) {
+    return useBinaryWriter((ByteDataWriter writer) {
+      writeSerializedUint64(writer, BigInt.parse(validityIntervalStart));
```
Comment (AlexDochioiu) [inline lib/src/utils/serialization_utils.dart:702]: Are you sure those `String` type numbers are not actually serialized as string in typescript? Cause it makes a huge difference.

File: lib/src/utils/serialization_utils.dart (hunk 3 above)
Comment (AlexDochioiu) [reply lib/src/utils/serialization_utils.dart:152]: Esti sigur ca acest ttl nu e serializat direct ca String in ts? 

De ex: Daca serializezi numarul 20 ca Uint sau "20" ca string o sa dea un rezultat complet diferit.

File: lib/src/utils/serialization_utils.dart (hunk 2 above)
Comment (AlexDochioiu) [reply lib/src/utils/serialization_utils.dart:145]: Esti sigur ca acest coin nu e serializat direct ca String in ts?

De ex: Daca serializezi numarul 20 ca Uint sau "20" ca string o sa dea un rezultat complet diferit.

File: lib/src/models/parsed_certificate.dart (hunk 1 above)
Comment (Kazopl) [reply lib/src/models/parsed_certificate.dart]: done

File: lib/src/utils/serialization_utils.dart (hunk 2 above)
Comment (Kazopl) [reply lib/src/utils/serialization_utils.dart:146]: in ts e `export type Uint64_str = string & {__type: 'uint64_t'}`  , el foloseste functia `export function serializeCoin(coin: Uint64_str): Buffer {
  return Buffer.concat([uint64_to_buf(coin)])
}` care converteste un string ce reprezinta un numar intreg pe 64 de biti intr-un Buffer

File: lib/src/utils/serialization_utils.dart (hunk 10 above)
Comment (Kazopl) [reply lib/src/utils/serialization_utils.dart:758]: clarified it in the latest commit

File: lib/src/models/parsed_certificate.dart [hunk 13]
```diff
@@ -8,69 +9,84 @@ part 'parsed_certificate.freezed.dart';
...
+  factory ParsedCertificate.authorizeCommitteeHot({
     required ParsedCredential coldCredential,
     required ParsedCredential hotCredential,
   }) = AuthorizeCommitteeHot;
 
-  const factory ParsedCertificate.resignCommitteeCold({
+  factory ParsedCertificate.resignCommitteeCold({
     required ParsedCredential coldCredential,
-     ParsedAnchor? anchor,
+    required ParsedAnchor anchor,
```
Comment (AlexDochioiu) [inline lib/src/models/parsed_certificate.dart:49]: Isn't `ParsedAnchor` nullable?

Comment (AlexDochioiu) [reply lib/src/models/parsed_certificate.dart:49]: Same to a few other factories in this file.

File: lib/src/utils/serialization_utils.dart (hunk 2 above)
Comment (AlexDochioiu) [reply lib/src/utils/serialization_utils.dart:146]: Hmm, ok. Atunci lasam implementarea in mare parte la fel. Dar `coin` tot trebuie schimbat in `BigInt` din `String`

File: lib/src/utils/serialization_utils.dart (hunk 3 above)
Comment (AlexDochioiu) [reply lib/src/utils/serialization_utils.dart:153]: Acelasi lucru aici. `ttl` ar trebui sa fie din start BigInt

File: lib/src/models/transaction_signing_mode.dart [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+  factory TransactionSigningModes.poolRegistrationAsOperator() = PoolRegistrationAsOperator;
+  factory TransactionSigningModes.plutusTransaction() = PlutusTransaction;
+  
+  late final int value = switch (this) {
+    OrdinaryTransaction() => TransactionSigningMode.ordinaryTransaction.value,
+    MultisigTransaction() => TransactionSigningMode.multisigTransaction.value,
+    PoolRegistrationAsOwner() => TransactionSigningMode.poolRegistrationAsOwner.value,
+    PoolRegistrationAsOperator() => TransactionSigningMode.poolRegistrationAsOperator.value,
+    PlutusTransaction() => TransactionSigningMode.plutusTransaction.value,
+  };
```
Comment (AlexDochioiu) [inline lib/src/models/transaction_signing_mode.dart:22]: Ma gandeam recent ca ar fi bine sa facem asta peste tot (+ sa stergem enum-ul complet si sa bagam valoarea direct aici). Am putea de asemenea sa redenumim acest field din `value` in `serializeTypeValue` ca sa fie mai specific.

Dar cred ca e mai bine sa facem modificarea asta dupa ce adaugam integration tests.

File: lib/src/models/version_compatibility.dart [hunk 15]
```diff
@@ -60,4 +69,168 @@ class VersionCompatibility with _$VersionCompatibility {
...
     );
   }
-}
\ No newline at end of file
+
+  static void ensureRequestSupportedByAppVersion(CardanoVersion version, ParsedSigningRequest request) {
+    final compatibility = VersionCompatibility.checkVersionCompatibility(version);
+
+    final void Function() invoker = switch (request.signingMode) {
+      PoolRegistrationAsOwner() when !compatibility.supportsPoolRegistrationAsOwner => () {
```
Comment (AlexDochioiu) [inline lib/src/models/version_compatibility.dart:77]: Nu am mai vazut pana acum `switch` cu `when` 😅 . Interesant.

**Pull request merged**